src/intrusive_pairing_heap.rs

```diff
@@ -0,0 +1,402 @@
+//! Implements an intrusive priority queue based on a pairing heap.
+
+use core::{
+    marker::PhantomPinned,
+    mem,
+    ops::{Deref, DerefMut, Drop},
+    ptr::NonNull,
+};
+
+/// Compares `a` and `b` without unwinding.
+/// This is necessary to avoid reentrancy in the heap.
+fn safe_lesser<T: Ord>(a: &T, b: &T) -> bool {
+    struct DropBomb;
+    impl Drop for DropBomb {
+        fn drop(&mut self) {
+            panic!("Panicked while comparing");
```

Comment:
Any benefit compared to aborting?

Reply:
Is there an easy way to abort in `no_std`?